cluster/service.go

```diff
@@ -51,74 +45,78 @@ type Stats struct {
 	CacheRefreshes int
 }
 
-// NewService creates a Resolver that rely on the Auth service to retrieve tokens
-func NewService(authURL string, clustersRefreshDelay time.Duration, serviceToken string, resolveToken token.Resolve, decode token.Decode, options ...configuration.HTTPClientOption) (Service, error) {
+type clusterService struct {
+	authService      *auth.Service
+	cacheRefresher   *time.Ticker
+	cacheRefreshLock *sync.RWMutex
+	cacheHits        int
+	cacheMissed      int
+	cacheRefreshes   int
+	cachedClusters   []Cluster
+}
+
+// NewClusterService creates an instance of service that using the Auth service retrieves information about clusters
+func NewClusterService(refreshInt time.Duration, authService *auth.Service) Service {
 	// setup a ticker to refresh the cluster cache at regular intervals
-	cacheRefresher := time.NewTicker(clustersRefreshDelay)
-	s := &clusterService{
-		authURL:          authURL,
-		serviceToken:     serviceToken,
-		resolveToken:     resolveToken,
-		decode:           decode,
+	cacheRefresher := time.NewTicker(refreshInt)
+	service := &clusterService{
+		authService:      authService,
 		cacheRefresher:   cacheRefresher,
 		cacheRefreshLock: &sync.RWMutex{},
-		clientOptions:    options}
-	// immediately load the list of clusters before returning
+	}
+	return service
+}
+
+func (s *clusterService) Start() error {
+	//immediately load the list of clusters before returning
 	err := s.refreshCache(context.Background())
 	if err != nil {
-		log.Error(nil, map[string]interface{}{"error_message": err}, "failed to load the list of clusters during service initialization")
-		return nil, err
+		return fmt.Errorf("failed to load the list of clusters during service initialization: %s", err)
 	}
 	go func() {
-		for range cacheRefresher.C { // while the `cacheRefresh` ticker is running
+		for range s.cacheRefresher.C { // while the `cacheRefresh` ticker is running
 			s.refreshCache(context.Background())
 		}
 	}()
-	return s, nil
+	return nil
```

Comment:
I don't see any benefit of adding separate `Start()` rather it add responsibility on consumer to call `Start()` without fail.  any specific reason for adding `Start()` ??

Also currently we are failing `NewClusterService()` if `refreshCache()` failed .. this is wrong thing .. like `refreshCache()` failed due to temporary network issue / service unavailable issue.

I guess, we can keep `NewClusterService()` as it is (without adding `Start()`) and not to fail when `refreshCache()` return error and hope that future call to `refreshCache()` will populate some cluster data. Cc: @alexeykazakov WDYT?

Reply:
> I don't see any benefit of adding separate Start() rather it add responsibility on consumer to call Start() without fail. any specific reason for adding Start()

there was a reason in the new implementation I took the code from - I was using it for local test purposes. Technically, I could work around it by setting a big refresh interval. So if you think that it doesn't make sense, then I can change it back to the original implementation and if I still need the `start()` method when I incorporate also the changes of the core logic, then I'll bring it back.

> Also currently we are failing NewClusterService() if refreshCache() failed .. this is wrong thing .. like refreshCache() failed due to temporary network issue / service unavailable issue.

this is what I was thinking about as well - if it should fail or not, but I kept the original implementation. Anyway, we need to have the tenant service ready once it's started, so ignoring the error is not the best idea I think. We can incorporate a [retry](https://github.com/MatousJobanek/fabric8-tenant/blob/v2/retry/retry.go) function that would try to get the list of clusters several times with some delay between the attempts. that would solve temporary network issues

Reply:
1. We need clusters to be loaded when starting a tenant service. Without having it the service is not functional. So, we the tenant service readiness prob should fail if we can't load clusters during startup. OpenShift will retry to re-start the pod a few times before giving up. If it's a temporal network issue then it will help. If not, then the old tenant service version with old cluster cache will continue to work. So no outages. Good :)

2. I don't see any issues with the new Start() method. NewClusterService() won't fail if cluster service is not available because it's not loading clusters but we call Start() in main too so the tenant service startup will fail on Start() which is good (see # 1 above)

3. Go routine started by Start() will keep working even if some attempts to refresh the cache fail. The service will be using the old cache until the network issue is gone. We just need to log the error here (you missed it and I've added a comment for that).

So, to summarize:
- Tenant service startup should fail if we can't load the initial list of clusters.
- If cache refresh fails then we need to continue our attempts to refresh it and keep using the old cache meanwhile.
It looks like this PR covers all of these above so we are good (just log the error when refresh fails).